lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -2476,10 +2430,87 @@ constructTransaction
+            guardOutsideOfValidityInterval validity mintingBurning
+            Right mintingBurning
+      where
+        handleMissingAssetName :: ApiMintBurnData n -> ApiMintBurnData n
+        handleMissingAssetName mb = case mb ^. #assetName of
+            Nothing -> mb {assetName = Just (ApiT nullTokenName)}
+            Just _ -> mb
+
+        guardWrongMintingTemplate
+            :: NonEmpty (ApiMintBurnData n) -> Either ErrConstructTx ()
+        guardWrongMintingTemplate mintingBurning =
+            when (any wrongMintingTemplate mintingBurning)
+                $ Left ErrConstructTxWrongMintingBurningTemplate
+          where
+            wrongMintingTemplate (ApiMintBurnData (ApiT script) _ _) =
+                isLeft (validateScriptOfTemplate RecommendedValidation script)
+                || countCosigners script /= (1 :: Int)
+                || existsNonZeroCosigner script
+            countCosigners = foldScript (const (+ 1)) 0
+            existsNonZeroCosigner =
+                foldScript (\cosigner a -> a || cosigner /= Cosigner 0) False
+
+        guardAssetNameTooLong
+            :: NonEmpty (ApiMintBurnData n) -> Either ErrConstructTx ()
+        guardAssetNameTooLong mintingBurning =
+            when (any assetNameTooLong mintingBurning)
+                $ Left ErrConstructTxAssetNameTooLong
+          where
+            assetNameTooLong = \case
+                ApiMintBurnData _ (Just (ApiT (UnsafeTokenName bs))) _ ->
+                    BS.length bs > tokenNameMaxLength
+                _ -> error "tokenName should be nonempty at this step"
+
+        guardAssetQuantityOutOfBounds
+            :: NonEmpty (ApiMintBurnData n) -> Either ErrConstructTx ()
+        guardAssetQuantityOutOfBounds mintingBurning =
+            when (any assetQuantityOutOfBounds mintingBurning)
+                $ Left ErrConstructTxMintOrBurnAssetQuantityOutOfBounds
+          where
+            assetQuantityOutOfBounds = \case
+                ApiMintBurnData _ _ (ApiMint (ApiMintData _ amt)) ->
+                    amt <= 0 || amt > unTokenQuantity txMintBurnMaxTokenQuantity
+                ApiMintBurnData _ _ (ApiBurn (ApiBurnData amt)) ->
+                    amt <= 0 || amt > unTokenQuantity txMintBurnMaxTokenQuantity
+
+        guardOutsideOfValidityInterval
+            :: (SlotNo, SlotNo)
+            -> NonEmpty (ApiMintBurnData n)
+            -> Either ErrConstructTx ()
+        guardOutsideOfValidityInterval (before, hereafter) mintingBurning =
+            when (any notWithinValidityInterval mintingBurning) $
+                Left ErrConstructTxValidityIntervalNotWithinScriptTimelock
+          where
+            notWithinValidityInterval (ApiMintBurnData (ApiT script) _ _) =
+                not $ withinSlotInterval before hereafter $
+                    scriptSlotIntervals script
+
+    validDelegationAction :: ApiMultiDelegationAction -> Bool
+    validDelegationAction = \case
```

Comment:
The word `valid` seems ambiguous here, since it's an adjective (though the type signature makes it clear).

Since we're returning a `Bool`, using the prefix `is` would make it less ambiguous, IMO. (Just like `isPrefixOf`, `isProperSubmapOf`, and so on.)

```suggestion
    isValidDelegationAction :: ApiMultiDelegationAction -> Bool
    isValidDelegationAction = \case
```

Reply:
You're right. 

I plan to continue the refactoring from this function, I want it to become a "parser-style" rather than a "predicate-style". We're missing a data type to represent valid-by-construction delegation action yet. Once its defined this expression could become:

``` 
delegationAction :: Handler DelegationAction
```

Reply:
or `parseDelegationAction :: Handler DelegationAction` if this is what sounds more straightforward to you (or maybe it's better because it doesn't occupy the name `delegationAction` which could be used down the path).